core/src/main/java/hudson/model/DirectoryBrowserSupport.java

```diff
@@ -436,33 +437,81 @@ private static void zip(StaplerResponse rsp, VirtualFile root, VirtualFile dir,
         try (ZipOutputStream zos = new ZipOutputStream(outputStream)) {
             zos.setEncoding(System.getProperty("file.encoding")); // TODO JENKINS-20663 make this overridable via query parameter
             // TODO consider using run(Callable) here
-            for (String n : dir.list(glob.isEmpty() ? "**" : glob, null, /* TODO what is the user expectation? */true)) {
-                String relativePath;
-                if (glob.length() == 0) {
-                    // JENKINS-19947: traditional behavior is to prepend the directory name
-                    relativePath = dir.getName() + '/' + n;
-                } else {
-                    relativePath = n;
-                }
 
-                String targetFile = dir.toString().substring(root.toString().length()) + n;
-                if (!ALLOW_SYMLINK_ESCAPE && root.supportIsDescendant() && !root.isDescendant(targetFile)) {
-                    LOGGER.log(Level.INFO, "Trying to access a file outside of the directory: " + root + ", illicit target: " + targetFile);
-                } else {
-                    // In ZIP archives "All slashes MUST be forward slashes" (http://pkware.com/documents/casestudies/APPNOTE.TXT)
-                    // TODO On Linux file names can contain backslashes which should not treated as file separators.
-                    //      Unfortunately, only the file separator char of the master is known (File.separatorChar)
-                    //      but not the file separator char of the (maybe remote) "dir".
-                    ZipEntry e = new ZipEntry(relativePath.replace('\\', '/'));
-                    VirtualFile f = dir.child(n);
-                    e.setTime(f.lastModified());
-                    zos.putNextEntry(e);
-                    try (InputStream in = f.open()) {
-                        IOUtils.copy(in, zos);
-                    }
-                    zos.closeEntry();
+            if (glob.isEmpty()) {
+                if (!root.supportsQuickRecursiveListing()) {
+                    // avoid slow listing when the Glob can do a quicker job
+                    glob = "**";
                 }
             }
+            
+            if (glob.isEmpty()) {
+                Map<String, VirtualFile> nameToVirtualFiles = collectRecursivelyAllLegalChildren(dir);
+                sendZipUsingMap(zos, dir, nameToVirtualFiles);
+            } else {
+                Collection<String> listOfFile = dir.list(glob, null, /* TODO what is the user expectation? */true);
+                sendZipUsingListOfNames(zos, dir, listOfFile);
+            }
+        }
+    }
+
+    private static void sendZipUsingMap(ZipOutputStream zos, VirtualFile dir, Map<String, VirtualFile> nameToVirtualFiles) throws IOException {
+        for (Map.Entry<String, VirtualFile> entry : nameToVirtualFiles.entrySet()) {
+            String n = entry.getKey();
+
+            String relativePath;
```

Comment:
Why not join declaration and assignment?

Reply:
Changed! good point, the code like this was more useful in the legacy version